Dont scroll to revealed files option.

phuein opened the pull request:
Adds a new option in settings to not scroll to revealed files in the file explorer view.

<!-- Thank you for submitting a Pull Request. Please: 
* Read our Pull Request guidelines:
  https://github.com/Microsoft/vscode/wiki/How-to-Contribute#pull-requests.
* Associate an issue with the Pull Request.
* Ensure that the code is up-to-date with the `master` branch.
* Include a description of the proposed changes and how to test them. 
-->

This PR fixes #23902

Quick demo: https://drive.google.com/file/d/1k_RL401kYfRgdfC1nVHkyPI65BSL-oXL/view?usp=sharing

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/microsoft/vscode?pullRequest=96890) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/vscode?pullRequest=96890) before we can accept your contribution.<br/><br/>:x: phuein [sign now](https://cla.opensource.microsoft.com/microsoft/vscode?pullRequest=96890)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/microsoft/vscode?pullRequest=96890) it.</sub>

isidorn commented on src/vs/workbench/contrib/files/browser/files.contribution.ts:
```diff
@@ -375,6 +375,11 @@ configurationRegistry.registerConfiguration({
 		},
+		'explorer.autoRevealNoScroll': {
+			'type': 'boolean',
+			'description': nls.localize('autoRevealNoScroll', "Controls whether the explorer should not automatically scroll to revealed files."),
```
Please chagne to "should automatically" the word "not" makes the sentence more confusing

isidorn commented on src/vs/workbench/contrib/files/browser/files.contribution.ts:
```diff
@@ -375,6 +375,11 @@ configurationRegistry.registerConfiguration({
 			'description': nls.localize('autoReveal', "Controls whether the explorer should automatically reveal and select files when opening them."),
 			'default': true
 		},
+		'explorer.autoRevealNoScroll': {
```
`explorer.autoRevealScroll` is a better name

clounie replied:
What about `explorer.autoScrollOnReveal` ?

As an end user that tells me exactly what's happening without even having to read the description.

clounie replied:
`autoScrollOnReveal` also highlights the dependency on `autoReveal` which you referenced here @phuein https://github.com/microsoft/vscode/pull/96890#discussion_r420491861

isidorn commented:
@phuein thank you for the PR.
I left some minor comments in your code.
Some things to think about: do users request for other auto reveal behavior and should this setting be a boolean or an open ended string. I suggest you search for all issues labeld with `file-explorer` and search for reveal. And see if there is more things which we could tackle as part of this setting.

However I am not 100% convinced we need this. The deafult behavior imho is fine
* If the file is visible do not scroll when auto revealing
* If explicitly explorer is focused do scroll

What is the case that does not work for you.

phuein commented:
Hi @isidorn, thanks for reviewing my PR.

I've modified the name and description to fit your suggestions. I agree it's less verbose and confusing this way.

I ran a search for `is:open label:file-explorer reveal` and I'm only seeing a repeat of this request / issue, so it doesn't seem like anything more needs to be added. If more requests come in relating to this issue, I'd be happy to review them, so feel free to tag me.

The case that bothers users is two-fold, and made us disable `explorer.autoReveal` entirely, which loses having the current file highlighted.

1. It may be unwanted for the explorer view to scroll away when closing a file tab. This requires the user to scroll back; repeatedly, when cleaning up tabs.

2. When moving between tabs often, the scrolling / changed project view may become a visual distraction.

isidorn commented:
Thanks for updating the PR and for claryfing.
However let's go deeper into 1 and 2.

1. So you have a.txt and b.txt open as tabs. Both are visible in the explorer. b.txt is focused, you close it and now a.txt is open. Explorer does not scroll for me. It only scroll if a.txt was not visible. Which is expected. Can you please clarify this?
2. Same as 1. when changing focus from a.txt to b.txt if both are visible no scrolling will happen.

clounie commented:
@isidorn have you read the original description? It is pretty detailed, this feels like a repeat discussion.

https://github.com/microsoft/vscode/issues/23902#issue-219382529

We just want to see the active file in the sidebar, without the sidebar scrolling on us when our active file changes.

clounie commented on src/vs/workbench/contrib/files/browser/files.contribution.ts:
```diff
@@ -375,6 +375,11 @@ configurationRegistry.registerConfiguration({
 		},
+		'explorer.autoRevealScroll': {
+			'type': 'boolean',
+			'description': nls.localize('autoRevealScroll', "Controls whether the explorer should automatically scroll to revealed files."),
```
Perhaps -
> Controls whether the explorer should automatically scroll when a different file is selected

phuein replied:
autoRevealScroll depends on autoReveal being active, so I feel the current description fits better.

phuein commented:
For both 1 and 2, this issue only refers to when the file is not visible in the explorer view, and so causing an unwanted scroll. With this in mind, my previous post explains scenarios where this scrolling may be unwanted.

Video example of scrolling scenario:
https://drive.google.com/file/d/1doXAwszkVNQtj2NUGDw0Fj8kwOkSz_tM/view?usp=sharing

isidorn commented:
@phuein so why don't you disable autoReveal? Because you want autoReval only when the item is in the viewport?

clounie commented:
Because if you disable autoReveal, then Explorer does not change the highlighted file.

clounie commented:
> @phuein so why don't you disable autoReveal? Because you want autoReval only when the item is in the viewport?

Because if you disable autoReveal, then Explorer does not correctly highlight the active file. It just keeps the highlight on the last-highlighted file until the user explicitly clicks on the correct file in Explorer.

clounie commented:
@isidorn here is a usecase (one of the main usecases) for this:

1. I'm in a Node.js project looking at a folder with 6 different files
2. I Cmd+Click to explore one of my dependencies in node_modules
3. VS Code proceeds to open that node module in explorer, opening a thousand different hierachies in the `node_modules` folder
4. Because of that incredibly annoying autoReveal behavior, I now have to scroll through hundreds if not _thousands_ of folders to close node modules so I can see the rest of my project hierarchy again

*That* is incredibly annoying as a user.

However, I still want to see the active file highlighted (just not scrolled to) because _that_ behavior is very useful.

isidorn commented:
Ok thanks for the explanation.
So what does not work for me in this PR is the setting name:
"autoRevealScroll"
If the setting is not scrolled into view it is not revealed, those are synonyms. Also it is funky that the autoRevealScroll depends on another setting autoReveal.

Idealy we would change the `autoReveal` setting to support both boolean values as before and string values.
So I can be `true | false | "on" | "off" | "highlightNoScroll"`.
That way all is under one setting and this feature would be more discoverable.

phuein commented:
I'm glad the issue and user needs are clarified. I also agree that the previous added setting was funky. I went with it as a simple concept. Modifying autoReveal added complexity. Especially with being boolean backwards compatible. It's now changed and implemented with the new commit.

Thanks for reviewing my work.

isidorn commented:
Thanks for the changes.
I tried it out and looks good overall.

However there is one problem, now when users who have explorer.autoReveal: true or false will see a warning when they open their settings.json
Unfortunetly we can not ship like this, since thousands of users might see this.
So we need to use a JSON Schema OR type. On the `file.contribution` line 374 change.

Apart from that:
1. You left a `console.log` in the code
2. I do not see the benefit of introducing the `AutoScroll` enum. That way you have to change the code in all the clients. Why not handle it locally in the explorerService so you expect true / false / string. That way changes will be local to that file.

![Screenshot 2020-05-11 at 11 28 48](https://user-images.githubusercontent.com/1926584/81546454-ad3d7280-937a-11ea-8517-6f818e8d2dd0.png)

phuein commented:
Alright, this commit should simplify it. It's either true / false / [string]. Less files and lines are affected. Also fixed the type error for the setting variable. And removed the excess console.log().

isidorn commented:
Thanks. However looking at the suggested values this is not good. It should suggest true false or the "highlightNoScroll" 
![Screenshot 2020-05-12 at 10 22 58](https://user-images.githubusercontent.com/1926584/81660068-9bbb9f80-943a-11ea-99bf-9962fd1fb41f.png)

isidorn commented on src/vs/workbench/contrib/files/browser/files.contribution.ts:
```diff
@@ -371,9 +371,15 @@ configurationRegistry.registerConfiguration({
+			'type': ['boolean', 'string'],
+			'enum': [true, false, 'highlightNoScroll'],
+			'default': 'autoReveal.on',
+			'enumDescriptions': [
```
I think this enum descriptions and the default are wrong. Default should be true.

phuein replied:
Nice catch, I missed that line. Committed again.

isidorn commented:
Good job, thanks for this PR. Merging in ☀️ 🍺

isidorn commented:
Pushed a bit of polish via a06c753e0b

phuein commented:
Awesome! Thanks for helping out throughout the PR. Everything should be checked now, and I can't wait to enjoy this on release.

isidorn commented:
Thanks for contributing!
